modality-probe-graph/src/lib.rs

```diff
@@ -0,0 +1,617 @@
+    tail_pending_edge_sources: HashMap<(ProbeId, SequenceNumber), GraphEvent>,
+
+    // This is to carry a clock-span across chunks.
+    last_event_by_probe_and_seq_num: HashMap<(ProbeId, SequenceNumber), GraphEvent>,
+
+    // This is the table used to look up the source events when a
+    // foreign clock is encountered in the log.
+    last_event_by_probe_and_clock: HashMap<(ProbeId, u32), GraphEvent>,
+}
+
+impl<G: Graph> EventDigraph<G> {
+    /// Construct an empty graph.
+    pub fn new(graph: G) -> Self {
+        EventDigraph {
+            graph,
+            tail_pending_edge_sources: HashMap::new(),
+            last_event_by_probe_and_seq_num: HashMap::new(),
+            last_event_by_probe_and_clock: HashMap::new(),
+        }
+    }
+
+    /// Turn a report into nodes and edges on the graph.
+    pub fn add_report(&mut self, report: &Report) -> Result<(), Error> {
+        let probe_id = report.probe_id;
+        let seq_num = report.seq_num;
+        let mut prev_event = None;
+        let mut first_event = true;
+        let mut self_clock = if let Some(sc) = report.frontier_clocks.get(0) {
+            sc
+        } else {
+            return Err(Error::InconsistentData("missing self frontier clock"));
+        };
+        let mut pending_edges = vec![];
+
+        for (idx, ev) in report.event_log.iter().enumerate() {
+            match ev {
+                EventLogEntry::Event(id) => {
+                    let node = GraphEvent {
+                        probe_id,
+                        id: *id,
+                        clock: *self_clock,
+                        payload: None,
+                        seq: seq_num,
+                        seq_idx: idx,
+                    };
+                    self.graph.add_node(node);
+                    if first_event {
+                        if let Some(tail) =
+                            self.tail_pending_edge_sources.remove(&(probe_id, seq_num))
+                        {
+                            self.graph.add_edge(tail, node);
+                        }
+                        if let Some(tail) = self
+                            .last_event_by_probe_and_seq_num
+                            .remove(&(probe_id, seq_num))
+                        {
+                            self.graph.add_edge(tail, node);
+                        }
+                        first_event = false;
+                    }
+                    if let Some(prev) = prev_event {
+                        self.graph.add_edge(prev, node);
+                    }
+                    for lc in pending_edges.iter() {
+                        if let Some(e) = self.last_event_by_probe_and_clock.get(lc) {
+                            self.graph.add_edge(*e, node);
+                        }
+                    }
+                    pending_edges.clear();
+                    prev_event = Some(node);
+                }
+                EventLogEntry::EventWithPayload(id, payload) => {
```

Comment:
Looks like the handling between Event and EventWithPayload is almost entirely duplicative, with the exception of intialization of the GraphEvent payload field. Non-essential, but it would be nice to collapse them down.